### PR TITLE
Improve the stability of GMLAN utils tests

### PR DESCRIPTION
**Pull request by polybassa**

### CODE REVIEW

File: test/contrib/automotive/gm/gmlanutils.uts [hunk 1]
```diff
@@ -158,15 +165,19 @@ with new_can_socket0() as isocan, ISOTPSocket(isocan, sid=0x242, did=0x642, base
...
 = Positive, pending message from different service interferes while pending
+drain_bus(iface0)
 started = threading.Event()
 def ecusim():
     with new_can_socket0() as isocan, ISOTPSocket(isocan, sid=0x642, did=0x242, basecls=GMLAN) as isotpsock2:
         isotpsock2.sniff(count=1, timeout=1, started_callback=started.set)
         pending = GMLAN()/GMLAN_NR(requestServiceId=0x34, returnCode=0x78)
         isotpsock2.send(pending)
         wrongservice = GMLAN()/GMLAN_NR(requestServiceId=0x36, returnCode=0x78)
+        time.sleep(0.1)
```
Comment (gpotter2) [inline test/contrib/automotive/gm/gmlanutils.uts:176]: What's this for? If you need to add delay for a test to pass, you likely need to refactor it...

**Review by gpotter2 — COMMENTED** [review]
Thanks for the PR.

While I appreciate the intent, adding delay in tests has been proven not to be stable. Do they really need it? This is not the way to go, you likely need to refactor those tests. If you're worried about their stability, we can always disable them in the meantime.

**Comment by polybassa** [comment]
Thank you very much for you feedback. 
I've analyzed this unit tests, and tried to understand why they are `randomly` failing in some corner cases. Unfortunately these tests are 100% stable on my local machine, so it's not possible for me, to get a local test setup to experiment with the instability. 
My analysis came to the following conclusion: Every of this test creates a couple of threads (at least: 1x ecusim, 2x ISOTPSoftSocket). The only thing, I can imagine, making some tests sometimes fail, are race conditions between these threads. 
I've identified three strategies to solve this:

1. Add time.sleep() to force a yield
2. Use `EcuAnsweringMachine` for the creation of responses
3. Some tests follow a certain procedure, add Events to synchronize the procedural steps between Ecu simulation and test-case 

The order of this strategies indicate the effort to apply it.    

Sidenote: 

Strategy 1: Forces a yield in the "ecusim" thread to ensure that the ISOTPSoftSocket threads get processing time in each procedural step of the test-cases. The factor 0.1 is chosen to ensure that even on windows systems the result of `time.now()` is not identical on two creations of a `Packet` in the ecusim thread. I assume, even a `time.sleep(0)` would already be sufficient, but first I wanted to exclude possible reasons for a bug. In case this changes make the tests stable, a change of time.sleep to the factor 0 would be possible.

Strategy 2: Would effectively be identical to strategy 1, since I've already added a `time.sleep` in the send procedure. In the past, this showed effectiveness in other unit tests. (https://github.com/secdev/scapy/blob/086fb5e5bff13f72c70638cdb1d903f2c2a9324d/scapy/contrib/automotive/ecu.py#L483) 

Strategy 3: This would require rewriting the test. I would like to try to solve the issue with Strategy 1 or 2, first.

I would be very happy, if you could merge this PR to apply this changes on the CI system. As soon as you discover instabilities in this tests, I propose to disable them until I find the time to rewrite them.

**Comment by gpotter2** [comment]
It's pretty unlucky: my first test rebased on this PR was https://github.com/secdev/scapy/pull/3090
It [froze on Ubuntu PyPy2](https://github.com/secdev/scapy/pull/3090/checks?check_run_id=2085230547) on the `gmlan` tests :/
```
[...]
passed 8A3A66FD 000.01s Craft Packet for NR
passed A5641895 000.01s Check not answers
passed 4B65CCE6 000.01s Check answers 1
passed 68DFC885 000.00s Set treat-response-pending-as-answer
passed 6AF7A292 000.01s Check response-pending is not considered as answer
passed 78C00978 000.01s Check response-pending is considered as answerSocketCAN support: False
Using PythonCANSocket virtual
TEST SKIPPED: can-isotp not available
Error: The operation was canceled.
```

**Comment by gpotter2** [comment]
I'll probably have to disable those tests. But you don't need to rewrite them all by yourself 😄

**Comment by gpotter2** [comment]
Hmm, this could be random. Maybe 20minutes is a bit short for the PyPy2 test and that's why it's failing

**Comment by polybassa** [comment]
Hi @gpotter2 

I've tried them with pypy in all versions on my local machine... unfortunately no errors, even if I let them run for hours (in a loop).

What about disabling this tests only on pypy, for now?

**Comment by gpotter2** [comment]
I tweaked the PyPy tests a big (made them run only as root) to try give the tests more runtime. (also we don't need to test it as non root).

Out of context info: I noticed the logs contain a lot of warnings/errors related to can sockets:
```
passed 05D89D30 000.00s Check UDS_SAPRif_nametoindex: No such device
SocketCAN setup done!
if_nametoindex: No such device
SocketCAN setup done!
SocketCAN support: True
Using NativeCANSocket
```

**Comment by polybassa** [comment]
These warnings are debug code to better understand the configuration of the test machine related to CAN. If you don't like them, I can easily remove them, but in case of a failed test, this allows to better understand the underlying can configuration of the system.

**Comment by polybassa** [comment]
Maybe this code caused the instability: https://github.com/secdev/scapy/pull/3075/commits/5940fa58413bb5b7fd597909f2bbf8a50eb080a9
I assume sprintf is quite heavy.

This code gets removed in #3078 and #3075

**Comment by gpotter2** [comment]
The instability could have actually just been a timeout.
I've tweaked PyPy timings https://github.com/secdev/scapy/pull/3143. We'll see how it goes
